### ISSUE TITLE
infer 时候 var not in this block

### ISSUE THREAD

**Issue by ws441572091** [PaddlePaddle/Paddle#12912]
fc层预测和训练都使用fc_tmp名字
crf层预测和训练时都使用crfw名字
crf层预测只有crf_decode

![image](https://user-images.githubusercontent.com/35720887/44563733-0ffb6e80-a792-11e8-837f-33ad6948f5e9.png)

预测程序不能直接使用crfw名字，可以使用fc_tmp参数
请问如何修改

![image](https://user-images.githubusercontent.com/35720887/44563750-302b2d80-a792-11e8-9cea-51ffe12dcb57.png)

**Comment by jacquesqiao** [comment]
模型是怎么配置的？

**Comment by ws441572091** [comment]
def train_program(word_dict,label_dict):
    crf_cost,crf_decode,target = infer_decode_program(word_dict,label_dict)
    avg_cost = fluid.layers.mean(crf_cost)
    p,r,f,ni,nl,nc = fluid.layers.chunk_eval(
            input = crf_decode,
            label = target,
            chunk_scheme = 'IOB',
            num_chunk_types = (len(label_dict)/2))

    return [avg_cost, f]


def optimizer_func():
    return fluid.optimizer.Adagrad(learning_rate=0.02)


def train(use_cuda,train_program, params_dirname):
    place = fluid.CUDAPlace(0) if use_cuda else fluid.CPUPlace()
    word_dict = paddle.dataset.conll03.word_dict()
    label_dict = paddle.dataset.conll03.label_dict()

    print("Reading training data....")
    train_reader = paddle.batch(paddle.reader.shuffle(
            paddle.dataset.conll03.train(word_dict,label_dict),buf_size=1000), batch_size=BATCH_SIZE)

    test_reader = paddle.batch(
            paddle.dataset.conll03.test(word_dict,label_dict),batch_size=BATCH_SIZE)

    valid_reader = paddle.batch(
            paddle.dataset.conll03.valid(word_dict,label_dict),batch_size=BATCH_SIZE)

    trainer = fluid.Trainer(
        train_func=partial(train_program, word_dict,label_dict),
        place=place,
        optimizer_func=optimizer_func)

    feed_order = ['words', 'label']

    tmp_acc =[0]
    def event_handler(event):
        if isinstance(event, fluid.EndStepEvent):
            if event.step % 200 == 0:

                print("Step {0}, Epoch {1} Metrics {2}".format(
                    event.step, event.epoch, map(np.array, event.metrics)))

        if isinstance(event, fluid.EndEpochEvent):
            avg_cost,acc = trainer.test(
                    reader = train_reader,feed_order = feed_order)
            print('Epoch {0}, Train Loss :{1},Acc {2:0.4}'.format(event.epoch, avg_cost,acc))
            avg_cost,acc = trainer.test(
                    reader = valid_reader,feed_order = feed_order)
            print('Epoch {0}, Valid Loss :{1},Acc {2:0.4}'.format(event.epoch, avg_cost,acc))
            print(tmp_acc)
            now_acc = tmp_acc[-1]
            if acc > now_acc:
                tmp_acc.append(acc)
                trainer.save_params(params_dirname)
                print ('Save params')
            avg_cost,acc = trainer.test(
                    reader = test_reader,feed_order = feed_order)
            print('Epoch {0}, Test Loss :{1},Acc {2:0.4}'.format(event.epoch, avg_cost,acc))


    trainer.train(
        num_epochs=1,
        event_handler=event_handler,
        reader=train_reader,
        feed_order=feed_order)


def infer(ues_cuda,inference_program,params_dirname=None,is_train=False):
    place = fluid.CUDAPlace(0) if use_cuda else fluid.CPUPlace()
    word_dict = paddle.dataset.conll03.word_dict()
    label_dict = paddle.dataset.conll03.label_dict()

    inferencer = fluid.Inferencer(
            infer_func = partial(inference_program,word_dict,label_dict,is_train),
            param_path = params_dirname,
            place = place)

    lod = []
    lod.append([17, 11983, 163, 1585, 165, 1340, 171, 37, 378, 36, 129, 2010, 421, 2, 3322, 3, 226, 46, 4648, 100, 1295, 4595, 9, 18, 6, 1351, 1273, 3, 2, 6259, 347, 205, 606, 221, 24, 36, 11979, 7354, 1, 4243, 0])
    base_shape = [[len(c) for c in lod]]

    tensor_words = fluid.create_lod_tensor(lod,base_shape,place)

    lod = []
    lod.append([0, 0, 0, 2, 2, 0, 0, 0, 0, 0, 3, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 4, 4, 4, 0, 0, 0, 0, 0, 0, 0, 1, 0])
    base_shape = [[len(c) for c in lod]]

    tensor_labels = fluid.create_lod_tensor(lod,base_shape,place)

    #results = inferencer.infer({'words':tensor_words},return_numpy = False)
    results = inferencer.infer({'words':tensor_words,},return_numpy = False)
    #results = crf_cost

    print([results[0].lod()])
    np_arr = np.array(results[0])
    np.set_printoptions(suppress=True)
    print(np_arr)

**Comment by ws441572091** [comment]
def train_program(word_dict,label_dict):
    crf_cost,crf_decode,target = infer_decode_program(word_dict,label_dict)
    avg_cost = fluid.layers.mean(crf_cost)
    p,r,f,ni,nl,nc = fluid.layers.chunk_eval(
            input = crf_decode,
            label = target,
            chunk_scheme = 'IOB',
            num_chunk_types = (len(label_dict)/2))

    return [avg_cost, f]


def optimizer_func():
    return fluid.optimizer.Adagrad(learning_rate=0.02)


def train(use_cuda,train_program, params_dirname):
    place = fluid.CUDAPlace(0) if use_cuda else fluid.CPUPlace()
    word_dict = paddle.dataset.conll03.word_dict()
    label_dict = paddle.dataset.conll03.label_dict()

    print("Reading training data....")
    train_reader = paddle.batch(paddle.reader.shuffle(
            paddle.dataset.conll03.train(word_dict,label_dict),buf_size=1000), batch_size=BATCH_SIZE)

    test_reader = paddle.batch(
            paddle.dataset.conll03.test(word_dict,label_dict),batch_size=BATCH_SIZE)

    valid_reader = paddle.batch(
            paddle.dataset.conll03.valid(word_dict,label_dict),batch_size=BATCH_SIZE)

    trainer = fluid.Trainer(
        train_func=partial(train_program, word_dict,label_dict),
        place=place,
        optimizer_func=optimizer_func)

    feed_order = ['words', 'label']

    tmp_acc =[0]
    def event_handler(event):
        if isinstance(event, fluid.EndStepEvent):
            if event.step % 200 == 0:

                print("Step {0}, Epoch {1} Metrics {2}".format(
                    event.step, event.epoch, map(np.array, event.metrics)))

        if isinstance(event, fluid.EndEpochEvent):
            avg_cost,acc = trainer.test(
                    reader = train_reader,feed_order = feed_order)
            print('Epoch {0}, Train Loss :{1},Acc {2:0.4}'.format(event.epoch, avg_cost,acc))
            avg_cost,acc = trainer.test(
                    reader = valid_reader,feed_order = feed_order)
            print('Epoch {0}, Valid Loss :{1},Acc {2:0.4}'.format(event.epoch, avg_cost,acc))
            print(tmp_acc)
            now_acc = tmp_acc[-1]
            if acc > now_acc:
                tmp_acc.append(acc)
                trainer.save_params(params_dirname)
                print ('Save params')
            avg_cost,acc = trainer.test(
                    reader = test_reader,feed_order = feed_order)
            print('Epoch {0}, Test Loss :{1},Acc {2:0.4}'.format(event.epoch, avg_cost,acc))


    trainer.train(
        num_epochs=1,
        event_handler=event_handler,
        reader=train_reader,
        feed_order=feed_order)


def infer(ues_cuda,inference_program,params_dirname=None,is_train=False):
    place = fluid.CUDAPlace(0) if use_cuda else fluid.CPUPlace()
    word_dict = paddle.dataset.conll03.word_dict()
    label_dict = paddle.dataset.conll03.label_dict()

    inferencer = fluid.Inferencer(
            infer_func = partial(inference_program,word_dict,label_dict,is_train),
            param_path = params_dirname,
            place = place)

    lod = []
    lod.append([17, 11983, 163, 1585, 165, 1340, 171, 37, 378, 36, 129, 2010, 421, 2, 3322, 3, 226, 46, 4648, 100, 1295, 4595, 9, 18, 6, 1351, 1273, 3, 2, 6259, 347, 205, 606, 221, 24, 36, 11979, 7354, 1, 4243, 0])
    base_shape = [[len(c) for c in lod]]

    tensor_words = fluid.create_lod_tensor(lod,base_shape,place)

    lod = []
    lod.append([0, 0, 0, 2, 2, 0, 0, 0, 0, 0, 3, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 4, 4, 4, 0, 0, 0, 0, 0, 0, 0, 1, 0])
    base_shape = [[len(c) for c in lod]]

    tensor_labels = fluid.create_lod_tensor(lod,base_shape,place)

    #results = inferencer.infer({'words':tensor_words},return_numpy = False)
    results = inferencer.infer({'words':tensor_words,},return_numpy = False)
    #results = crf_cost

    print([results[0].lod()])
    np_arr = np.array(results[0])
    np.set_printoptions(suppress=True)
    print(np_arr)

**Comment by ws441572091** [comment]
def train_program(word_dict,label_dict):
    crf_cost,crf_decode,target = infer_decode_program(word_dict,label_dict)
    avg_cost = fluid.layers.mean(crf_cost)
    p,r,f,ni,nl,nc = fluid.layers.chunk_eval(
            input = crf_decode,
            label = target,
            chunk_scheme = 'IOB',
            num_chunk_types = (len(label_dict)/2))

    return [avg_cost, f]

**Comment by ws441572091** [comment]
![image](https://user-images.githubusercontent.com/35720887/44568878-18ac6e80-a7ab-11e8-929b-bb0d3022bb3c.png)
![image](https://user-images.githubusercontent.com/35720887/44568904-2b26a800-a7ab-11e8-9975-c38a867ce25f.png)
![image](https://user-images.githubusercontent.com/35720887/44568954-64f7ae80-a7ab-11e8-9fed-272fed1e4336.png)

**Comment by jacquesqiao** [comment]
请不要使用inferencer接口，目前缺乏维护，建议使用官方支持的api。
